### DISCUSSION TITLE
Remove `@types/node` from default `public-hoist-pattern`?

### DISCUSSION THREAD

**Discussion by loynoir** [pnpm/pnpm#4455] [Q&A] [answered]
### Feature Request
Remove `@types/node` from default `public-hoist-pattern`?

But I guess it might be too aggressive strategy, so open a QA first.

### Related issue
https://github.com/microsoft/TypeScript/issues/48341

### Three cases
Brief, given
- tsconfig have a limited target, eg: es2016
- code have es2017

tsc complain es2017 as expected

---
Given
- tsconfig have a limited target, eg: es2016
- code have es2017
- install `sort-package-json@1.54.0`, which somehow includes `@types/node@17.0.21`

tsc will not complain es2017

---
Given
- tsconfig have a limited target, eg: es2016
- code have es2017
- install `sort-package-json@1.54.0`, which somehow includes `@types/node@17.0.21`
- disallow hoist `@types/node@17.0.21`, eg: exclude types from default
```
; https://pnpm.io/npmrc#public-hoist-pattern
; remove '*types*' from default
; public-hoist-pattern[]=*types*
public-hoist-pattern[]=*eslint*
public-hoist-pattern[]=@prettier/plugin-*
public-hoist-pattern[]=*prettier-plugin-*
```

tsc complain es2017 as expected

**Comment by zkochan** [comment, marked as answer]
I think it might sense to remove hoisting types.

**Reply by loynoir** [reply]
1. Is this a valid format to write into .npmrc?
```
public-hoist-pattern[]= '*eslint*', '@prettier/plugin-*', '*prettier-plugin-*
```

2. Can I write something like this? 
```
public-hoist-pattern[] = default, '!*@types/node*'
```

3. Is it safe to not hoist `@types`?

**Reply by zkochan** [reply]
1

```
public-hoist-pattern[]=*eslint*
public-hoist-pattern[]=@prettier/plugin-*
public-hoist-pattern[]=*prettier-plugin-*
```

2 no

3 yes

**Comment by Jack-Works** [comment]
Good idea. I have another suggesion, default to force @types/node to the same version (do not allow multiple @types/node version in the dependency, this will make typescript unhappy)

**Reply by jekh** [reply]
I added this to package.json to get rid of the multiple conflicting @types/node:
```json
  "pnpm": {
    "overrides": {
      "@types/node": "^16"
    }
  }
  ```

**Reply by loynoir** [reply]
@jekh https://github.com/pnpm/pnpm/pull/4459
